// @medplum/react/dist/esm/ServiceRequestTimeline/ServiceRequestTimeline.mjs
import { getReferenceString, createReference } from '@medplum/core';
import React from 'react';
import { ResourceTimeline } from '../ResourceTimeline/ResourceTimeline.mjs';

function ServiceRequestTimeline(props) {
    return (React.createElement(ResourceTimeline, { value: props.serviceRequest, buildSearchRequests: (resource) => ({
            resourceType: 'Bundle',
            type: 'batch',
            entry: [
                {
                    request: {
                        method: 'GET',
                        url: `${getReferenceString(resource)}/_history`,
                    },
                },
                {
                    request: {
                        method: 'GET',
                        url: `Communication?based-on=${getReferenceString(resource)}&_sort=-_lastUpdated`,
                    },
                },
                {
                    request: {
                        method: 'GET',
                        url: `Media?_count=100&based-on=${getReferenceString(resource)}&_sort=-_lastUpdated`,
                    },
                },
                {
                    request: {
                        method: 'GET',
                        url: `DiagnosticReport?based-on=${getReferenceString(resource)}&_sort=-_lastUpdated`,
                    },
                },
            ],
        }), createCommunication: (resource, sender, text) => ({
            resourceType: 'Communication',
            status: 'completed',
            basedOn: [createReference(resource)],
            subject: resource.subject,
            sender: createReference(sender),
            sent: new Date().toISOString(),
            payload: [{ contentString: text }],
        }), createMedia: (resource, operator, content) => ({
            resourceType: 'Media',
            status: 'completed',
            basedOn: [createReference(resource)],
            subject: resource.subject,
            operator: createReference(operator),
            issued: new Date().toISOString(),
            content,
        }) }));
}

export { ServiceRequestTimeline };
