// @medplum/react/dist/esm/utils/questionnaire.mjs
var QuestionnaireItemType;
(function (QuestionnaireItemType) {
    QuestionnaireItemType["group"] = "group";
    QuestionnaireItemType["display"] = "display";
    QuestionnaireItemType["question"] = "question";
    QuestionnaireItemType["boolean"] = "boolean";
    QuestionnaireItemType["decimal"] = "decimal";
    QuestionnaireItemType["integer"] = "integer";
    QuestionnaireItemType["date"] = "date";
    QuestionnaireItemType["dateTime"] = "dateTime";
    QuestionnaireItemType["time"] = "time";
    QuestionnaireItemType["string"] = "string";
    QuestionnaireItemType["text"] = "text";
    QuestionnaireItemType["url"] = "url";
    QuestionnaireItemType["choice"] = "choice";
    QuestionnaireItemType["openChoice"] = "open-choice";
    QuestionnaireItemType["attachment"] = "attachment";
    QuestionnaireItemType["reference"] = "reference";
    QuestionnaireItemType["quantity"] = "quantity";
})(QuestionnaireItemType || (QuestionnaireItemType = {}));
function isChoiceQuestion(item) {
    return item.type === 'choice' || item.type === 'open-choice';
}

export { QuestionnaireItemType, isChoiceQuestion };
