// @medplum/react/dist/esm/auth/NewUserForm.mjs
import { Center, Alert, Group, Divider, Stack, TextInput, PasswordInput, Text, Anchor, Checkbox, Button } from '@mantine/core';
import { IconAlertCircle } from '@tabler/icons';
import React, { useState, useEffect } from 'react';
import { Form } from '../Form/Form.mjs';
import { getGoogleClientId, GoogleButton } from '../GoogleButton/GoogleButton.mjs';
import { useMedplum } from '../MedplumProvider/MedplumProvider.mjs';
import { getIssuesForExpression, getErrorsForInput } from '../utils/outcomes.mjs';
import { initRecaptcha, getRecaptcha } from '../utils/recaptcha.mjs';

function NewUserForm(props) {
    const googleClientId = getGoogleClientId(props.googleClientId);
    const recaptchaSiteKey = props.recaptchaSiteKey;
    const medplum = useMedplum();
    const [outcome, setOutcome] = useState();
    const issues = getIssuesForExpression(outcome, undefined);
    useEffect(() => initRecaptcha(recaptchaSiteKey), [recaptchaSiteKey]);
    return (React.createElement(Form, { style: { maxWidth: 400 }, onSubmit: async (formData) => {
            try {
                const recaptchaToken = await getRecaptcha(recaptchaSiteKey);
                props.handleAuthResponse(await medplum.startNewUser({
                    projectId: props.projectId,
                    firstName: formData.firstName,
                    lastName: formData.lastName,
                    email: formData.email,
                    password: formData.password,
                    remember: formData.remember === 'true',
                    recaptchaSiteKey,
                    recaptchaToken,
                }));
            }
            catch (err) {
                setOutcome(err);
            }
        } },
        React.createElement(Center, { sx: { flexDirection: 'column' } }, props.children),
        issues && (React.createElement(Alert, { icon: React.createElement(IconAlertCircle, { size: 16 }), color: "red" }, issues.map((issue) => (React.createElement("div", { "data-testid": "text-field-error", key: issue.details?.text }, issue.details?.text))))),
        googleClientId && (React.createElement(React.Fragment, null,
            React.createElement(Group, { position: "center", p: "xl", style: { height: 70 } },
                React.createElement(GoogleButton, { googleClientId: googleClientId, handleGoogleCredential: async (response) => {
                        try {
                            await medplum.startPkce();
                            props.handleAuthResponse(await medplum.startGoogleLogin({
                                googleClientId: response.clientId,
                                googleCredential: response.credential,
                                createUser: true,
                            }));
                        }
                        catch (err) {
                            setOutcome(err);
                        }
                    } })),
            React.createElement(Divider, { label: "or", labelPosition: "center", my: "lg" }))),
        React.createElement(Stack, { spacing: "xl" },
            React.createElement(TextInput, { name: "firstName", type: "text", label: "First name", placeholder: "First name", required: true, autoFocus: true, error: getErrorsForInput(outcome, 'firstName') }),
            React.createElement(TextInput, { name: "lastName", type: "text", label: "Last name", placeholder: "Last name", required: true, error: getErrorsForInput(outcome, 'lastName') }),
            React.createElement(TextInput, { name: "email", type: "email", label: "Email", placeholder: "name@domain.com", required: true, error: getErrorsForInput(outcome, 'email') }),
            React.createElement(PasswordInput, { name: "password", label: "Password", autoComplete: "off", required: true, error: getErrorsForInput(outcome, 'password') }),
            React.createElement(Text, { color: "dimmed", size: "xs" },
                "By clicking submit you agree to the Medplum",
                ' ',
                React.createElement(Anchor, { href: "https://www.medplum.com/privacy" }, "Privacy\u00A0Policy"),
                ' and ',
                React.createElement(Anchor, { href: "https://www.medplum.com/terms" }, "Terms\u00A0of\u00A0Service"),
                "."),
            React.createElement(Text, { color: "dimmed", size: "xs" },
                "This site is protected by reCAPTCHA and the Google",
                ' ',
                React.createElement(Anchor, { href: "https://policies.google.com/privacy" }, "Privacy\u00A0Policy"),
                ' and ',
                React.createElement(Anchor, { href: "https://policies.google.com/terms" }, "Terms\u00A0of\u00A0Service"),
                " apply.")),
        React.createElement(Group, { position: "apart", mt: "xl", noWrap: true },
            React.createElement(Checkbox, { name: "remember", label: "Remember me", size: "xs" }),
            React.createElement(Button, { type: "submit" }, "Create account"))));
}

export { NewUserForm };
