// @medplum/react/dist/esm/DescriptionList/DescriptionList.mjs
import { createStyles } from '@mantine/core';
import React from 'react';

const useStyles = createStyles((theme) => ({
    root: {
        display: 'grid',
        gridTemplateColumns: '30% 70%',
        margin: 0,
        '& > dt, & > dd': {
            padding: `${theme.spacing.sm}px ${theme.spacing.sm}px`,
            borderTop: `0.1px solid ${theme.colors.gray[3]}`,
            margin: 0,
        },
    },
    compact: {
        gridTemplateColumns: '20% 80%',
        '& > dt, & > dd': {
            padding: `0 ${theme.spacing.xs}px ${theme.spacing.xs}px 0`,
            border: 0,
        },
    },
}));
function DescriptionList(props) {
    const { children, compact } = props;
    const { classes, cx } = useStyles();
    return React.createElement("dl", { className: cx(classes.root, { [classes.compact]: compact }) }, children);
}
function DescriptionListEntry(props) {
    return (React.createElement(React.Fragment, null,
        React.createElement("dt", null, props.term),
        React.createElement("dd", null, props.children)));
}

export { DescriptionList, DescriptionListEntry };
