// @medplum/react/dist/esm/Panel/Panel.mjs
import { createStyles, useComponentDefaultProps, Paper } from '@mantine/core';
import React from 'react';

const useStyles = createStyles((theme, { width, fill }) => ({
    paper: {
        maxWidth: width,
        margin: `${theme.spacing.xl}px auto`,
        padding: fill ? 0 : theme.spacing.md,
        '@media (max-width: 800px)': {
            padding: fill ? 0 : 8,
        },
        '& img': {
            width: '100%',
            maxWidth: '100%',
        },
        '& video': {
            width: '100%',
            maxWidth: '100%',
        },
    },
}));
const defaultProps = {
    shadow: 'xs',
    radius: 'md',
    withBorder: true,
};
function Panel(props) {
    const { className, children, width, fill, unstyled, ...others } = useComponentDefaultProps('Panel', defaultProps, props);
    const { classes, cx } = useStyles({ width, fill }, { name: 'Panel', unstyled });
    return (React.createElement(Paper, { className: cx(classes.paper, className), ...others }, children));
}

export { Panel };
